aimon/reprompting_api/pipeline.py

```diff
@@ -0,0 +1,544 @@
+
+        # Select best response across all iterations
+        best_output, best_failed_count = self._select_best_iteration(iteration_outputs)
+            
+        # Build final response payload
+        response = {"best_response": best_output}
+        if self.config.return_telemetry:
+            response["telemetry"] = self.telemetry.get_all()
+        if self.config.return_aimon_summary:
+            response["summary"] = self._gen_summary(iteration_num, best_failed_count)
+            
+        logger.info("RepromptingPipeline run completed")
+        logger.info(f"Best response selected with {best_failed_count} failed instructions remaining.")
+
+        return response  
+
+    def _build_original_prompt(self) -> Template:
+        """
+        Build a reusable template for combining system_prompt, context, and user_query.
+        This returns a string.Template object so the caller can safely substitute values.
+        
+        Placeholders:
+            - system_prompt
+            - context
+            - user_query
+
+        Returns:
+            Template: A string.Template for building the base LLM prompt.
+        """
+        template_str = (
+            "System:\n${system_prompt}\n\n"
+            "Context:\n${context}\n\n"
+            "User Query:\n${user_query}"
+        )
+        return Template(template_str)
+
+    def _build_aimon_payload(self, context, user_query, user_instructions, generated_text, system_prompt):
+        """
+        Constructs AIMon input payload.
+
+        Args:
+            context (str): Context for the LLM.
+            user_query (str): The user's query.
+            user_instructions (list[str]): Instructions for the model.
+            generated_text (str): The model's generated response.
+
+        Returns:
+            dict: Payload for AIMon evaluation.
+        """
+        if not isinstance(user_instructions, list):
+            user_instructions = []
+        payload = {
+            'context': context,
+            'user_query': user_query,
+            'generated_text': generated_text,
+            'instructions': user_instructions,
+            'system_prompt' : system_prompt
+        }
+        return payload
+
+    def _call_llm(self, prompt_template: Template, max_attempts, system_prompt=None, context=None, user_query=None, base_delay=1):
+        """
+        Calls the LLM with exponential backoff. Retries if the LLM call fails
+        OR returns a non-string value. Raises an exception if all retries fail.
+
+        Args:
+            prompt_template (Template): Prompt template for the LLM.
+            max_attempts (int): Max retry attempts.
+            base_delay (float): Initial delay in seconds before backoff.
+
+        Returns:
+            str: LLM response text.
+
+        Raises:
+            RuntimeError: If the LLM call fails or returns an invalid type after all retries.
+        """
+        last_exception = None
+        for attempt in range(max_attempts):
```

Comment:
lets create a utility in the utils.py for exponential backoff. Here is a retry decorator that you can use: https://gist.github.com/pjoshi30/74c177130aff043fa5bfde4d4cac54d2